contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -4029,3 +4035,309 @@ exec_stmt_fulltextindex(PLtsql_execstate *estate, PLtsql_stmt_fulltextindex *stm
 
 	return PLTSQL_RC_OK;
 }
+
+/*
+ * tsql_compare_values
+ *		Note: This function is used to sort the values in the array.
+ *		It compare two datum values using the function oid of operator provided in arg,
+ *		it also sets the contains_duplicate flag in the context if duplicate
+ *		values are found. This is modified version of compare_values().
+ *		Returns -1 if a < b, 1 if a > b and 0 if a == b.
+ */
+static int
+tsql_compare_values(const void *a, const void *b, void *arg)
+{
+	Datum		*da = (Datum *) a;
+	Datum		*db = (Datum *) b;
+	Datum		r;
+
+	tsql_compare_context *cxt = (tsql_compare_context *) arg;
+
+	r = OidFunctionCall2Coll(cxt->function_oid, cxt->colloid, *da, *db);
+
+	if (DatumGetBool(r))
+		return -1;
+
+	r = OidFunctionCall2Coll(cxt->function_oid, cxt->colloid, *db, *da);
+
+	if (DatumGetBool(r))
+		return 1;
+
+	cxt->contains_duplicate = true;
+	return 0;
+}
+
+/*
+ * Allows only login that is either db owner or member of sysadmin.
+ */
+static void 
+perform_permission_check(char *name, bool is_create, bool is_function)
+{
+	char		*dbname = get_cur_db_name();
+	Oid		session_user_id = GetSessionUserId();
+	char		*login = GetUserNameFromId(session_user_id, false);
+	bool		login_is_db_owner;
+
+	login_is_db_owner = 0 == strncmp(login, get_owner_of_db(dbname), NAMEDATALEN);
+
+	if (!login_is_db_owner && !is_member_of_role(session_user_id, get_role_oid("sysadmin", false)))
+	{
+		if(is_create)
+			ereport(ERROR, 
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE), 
+					errmsg("User does not have permission to perform this action.")));
+		else
+			ereport(ERROR, 
+				(errcode(ERRCODE_UNDEFINED_OBJECT), 
+					errmsg("Cannot drop the partition %s '%s', because it does not exist or you do not have permission.", 
+							(is_function? "function": "scheme"), name)));
+	}
+}
+
+/*
+ * CREATE/DROP PARTITION FUNCTION
+ */
+static int
+exec_stmt_partition_function(PLtsql_execstate *estate, PLtsql_stmt_partition_function *stmt)
+{
+	char *partition_function_name = stmt->function_name;
+
+	/* check if the login has necessary permissions for CREATE/DROP */
+	perform_permission_check(partition_function_name, stmt->is_create, true);
+
+	if (stmt->is_create) 
+	{
+		PLtsql_type		*typ = stmt->datatype;
+		List 			*arg = stmt->args;
+		bool 			isnull;
+		Oid			valtype;
+		int32			valtypmod;
+		Datum			tsql_type_datum;
+		char			*tsql_typename = NULL;
+		Datum			*values = NULL;
+		ArrayType		*arr_value = NULL;
+		Oid			sql_variant_oid;
+		tsql_compare_context 	cxt;
+		Operator		operator;
+		Oid			oprcode;
+		int 			nargs = list_length(arg);
+		MemoryContext		cur_ctxt = CurrentMemoryContext;
+		LOCAL_FCINFO(fcinfo, 1);
+
+		/* check if given name is exceeding the allowed limit */
+		if(strlen(partition_function_name) > 128)
+		{
+			ereport(ERROR, 
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("The identifier that starts with '%.128s' is too long. Maximum length is 128.", partition_function_name)));
+		}
```

Comment:
I dont think we will ever hit this here. Have we added any test which executes this code?